lasy/profiles/from_openpmd_profile.py

```diff
@@ -0,0 +1,129 @@
+
+    pol : list of 2 complex numbers (dimensionless)
+        Polarization vector. It corresponds to :math:`p_u` in the above
+        formula ; :math:`p_x` is the first element of the list and
+        :math:`p_y` is the second element of the list. Using complex
+        numbers enables elliptical polarizations.
+
+    field : string
+        Name of the field containing the laser pulse
+        Passed directly OpenPMDTimeSeries.
+
+    coord : string
+        Name of the field containing the laser pulse
+        Passed directly OpenPMDTimeSeries.
+
+    envelope : boolean
+        Whether the file represents a laser envelope.
+        If not, the envelope is obtained from the electric field
+        using a Hilbert transform
+
+    prefix : string
+        Prefix of the openPMD file from which the envelope is red.
+        Only used when envelope=True.
+        The provided iteration is read from <path>/<prefix>_%T.h5.
+
+    wavelength : float (in meter)
+        The main laser wavelength :math:`\\lambda_0` of the laser, which defines
+        :math:`\\omega_0`, according to :math:`\\omega_0 = 2\\pi c/\\lambda_0`.
+        This argument is optional, and will overwrite the following default
+        behavior:
+        - if envelope is True, the central wavelength is read from the openPMD
+        file at openPMD envelope standard.
+        - if envelope is False, the central wavelength is obtained from the
+        Hilbert transform.
+    """
+
+    def __init__(
+        self,
+        path,
+        iteration,
+        pol,
+        field,
+        coord=None,
+        envelope=False,
+        prefix=None,
+        wavelength=None,
+    ):
+        ts = OpenPMDTimeSeries(path)
+        F, m = ts.get_field(iteration=iteration, field=field, coord=coord, theta=None)
+        assert m.axes in [
+            {0: "x", 1: "y", 2: "z"},
+            {0: "z", 1: "y", 2: "x"},
+            {0: "x", 1: "y", 2: "t"},
+            {0: "t", 1: "y", 2: "x"},
+        ]
+
+        if m.axes in [{0: "z", 1: "y", 2: "x"}, {0: "t", 1: "y", 2: "x"}]:
+            F = F.swapaxes(0, 2)
+
+        if "z" in m.axes.values():
+            dt = m.dz / c
+            t = (m.z - m.z[0]) / c
+        else:
+            dt = m.dt
+            t = m.t
+        axes = {"x": m.x, "y": m.y, "t": t}
+
+        # If array does not contain the envelope but the electric field,
+        # extract the envelope with a Hilbert transform
+        if envelope == False:
+            # Assumes z is last dimension!
+            h = hilbert(F)
+
+            if "z" in m.axes.values():
+                # Flip to get complex envelope in t assuming z = -c*t
+                h = np.flip(h, axis=2)
+
+            # Get central wavelength from array
+            phase = np.unwrap(np.angle(h))
+            omg0_h = (
+                -np.average(
+                    np.diff(phase),
+                    weights=0.5 * (np.abs(h[:, :, 1:]) + np.abs(h[:, :, :-1])),
```

Comment:
here you could probably use `np.gradient(phase, axis=-1)` with `weights=np.abs(h)`